sympy/calculus/range_func.py

```diff
@@ -0,0 +1,170 @@
+
+    Parameters
+    ==========
+
+    func: Expr
+          The expression whose range is to be found
+    set_value: Union of Sets
+          The domain for the variable involved in function
+
+    Raises
+    ======
+
+    NotImplementedError
+          The algorithms for to find the solution of the given equation are
+          not yet implemented.
+    ValueError
+          The input is not valid.
+    RuntimeError
+          It is a bug, please report to the github issue tracker.
+
+    Examples
+    ========
+
+    >>> from sympy.calculus.range_func import range_func
+    >>> from sympy import Symbol, S, Interval, Union, FiniteSet
+    >>> x = Symbol('x', real=True)
+    >>> range_func(x**2, Interval(-1, 1))
+    [0, 1]
+    >>> range_func(x/(x**2 - 4), Union(Interval(-1, 3), FiniteSet(5)))
+    (-oo, 1/3] U [3/5, oo)
+    >>> range_func(x**2/(x**2 - 4), S.Reals)
+    (-oo, 0] U (1, oo)
+    """
+
+    func = sympify(func)
+    if not isinstance(func, (Expr, Number)):
+        raise ValueError(" %s is not a valid sympy expression" % (func))
+
+    if not isinstance(set_value, Set):
+        raise ValueError('A Symbol must be given, not type %s: %s' % (type(set_value), set_value))
+
+    if set_value.is_EmptySet:
+        return EmptySet()
+
+    free_symbol = func.free_symbols
+    if len(free_symbol) == 1:
+        symbol = free_symbol.pop()
+    elif len(free_symbol) == 0:
+        return FiniteSet(func)
+    else:
+        raise NotImplementedError("more than one variables %s not handeled" % (free_symbol))
+
+    if not func.is_rational_function(symbol):
+        raise NotImplementedError("Algorithms finding range for non-rational functions"
+                                    "are not yet implemented")
+
+    # this block of code can be replaced by
+    # sing = Intersection(FiniteSet(*sing), set_val.closure)
+    # after the issue #9706 has been fixed
+    def closure_handle(set_im, singul):
+        if set_im.has(Interval):
+            if not oo in set_im.boundary:
+                if not S.NegativeInfinity in set_im.boundary:
+                    return Intersection(FiniteSet(*singul), set_im.closure)
+                return Intersection(FiniteSet(*singul), Union(set_im,
+                                    FiniteSet(max(set_im.boundary))))
+            else:
+                if not S.NegativeInfinity in set_im.boundary:
+                    return Intersection(FiniteSet(*singul), Union(set_im,
+                                        FiniteSet(min(set_im.boundary))))
+                return Intersection(FiniteSet(*singul), set_im)
+        return Intersection(FiniteSet(*singul), set_im)
+
+    # all the singularities of the function
+    sing = singularities(func, symbol)
+    sing = closure_handle(set_value, sing)
+
+    def in_intrvl(f, set_val):
+        val1 = (set_val.start, set_val.args[2])
+        val2 = (set_val.end, set_val.args[3])
+        g1 = simplify(diff(f, symbol))
```

Comment:
`g1` is a very uninformative name better use `df` and for `g2` below you can use `d2f`.